Mathlib/LinearAlgebra/RootSystem/Basic.lean

```diff
@@ -0,0 +1,336 @@
+ * `RootPairing.IsCrystallographic`: A root pairing is said to be crystallographic if the pairing
+   between a root and coroot is always an integer.
+ * `RootPairing.IsReduced`: A root pairing is said to be reduced if it never contains the double of
+   a root.
+ * `RootPairing.ext`: In characteristic zero if there is no torsion, the correspondence between
+   roots and coroots is unique.
+ * `RootSystem.ext`: In characteristic zero if there is no torsion, a root system is determined
+   entirely by its roots.
+ * `RootSystem.mk'`: In characteristic zero if there is no torsion, to check that a family of
+   roots form a root system, we do not need to check that the coroots are stable under reflections
+   since this follows from the corresponding property for the roots.
+
+## Implementation details
+
+A root datum is sometimes defined as two subsets: roots and coroots, together with a bijection
+between them, subject to hypotheses. However the hypotheses ensure that the bijection is unique and
+so the question arises of whether this bijection should be part of the data of a root datum or
+whether one should merely assert its existence. For root systems, things are even more extreme: the
+coroots are uniquely determined by the roots. Furthermore a root system induces a canonical
+non-degenerate bilinear form on the ambient space and many informal accounts even include this form
+as part of the data.
+
+We have opted for a design in which some of the uniquely-determined data is included: the bijection
+between roots and coroots is (implicitly) included and the coroots are included for root systems.
+Empirically this seems to be by far the most convenient design and by providing extensionality
+lemmas expressing the uniqueness we expect to get the best of both worlds.
+
+A final point is that we require roots and coroots to be injections from a base indexing type `ι`
+rather than subsets of their codomains. This design was chosen to avoid the bijection between roots
+and coroots being a dependently-typed object. A third option would be to have the roots and coroots
+be subsets but to avoid having a dependently-typed bijection by defining it globally with junk value
+`0` outside of the roots and coroots. This would work but lacks the convenient symmetry that the
+chosen design enjoys: by introducing the indexing type `ι`, one does not have to pick a direction
+(`roots → coroots` or `coroots → roots`) for the forward direction of the bijection. Besides,
+providing the user with the additional definitional power to specify an indexing type `ι` is a
+benefit and the junk-value pattern is a cost.
+
+-/
+
+open Set Function
+open Module hiding reflection
+open Submodule (span)
+open AddSubgroup (zmultiples)
+
+noncomputable section
+
+variable (ι : Type*) [Finite ι] {R M N : Type*}
+  [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+  [IsReflexive R M] (e : N ≃ₗ[R] Dual R M)
```

Comment:
The setup seems to break symmetry between `M` and `N`. Because only `M` is required to be reflexive.
Can you prove `RootPairing.flip` which takes a root pairing on `(M, N)` and turns it into a root pairing on `(N, M)` by exchanging the roles of the roots and coroots?

My intuition suggests to encode `e` as a bilinear map
```
e : M →ₗ[R] N →ₗ[R] R
```
so that you can use `e.flip` to swap the roles of `M` and `N`.
Of course you would then need to require the pairing to be perfect, and I think this could be done as one of the fields of `RootPairing`.
There is of course `LinearEquiv.flip`, but it just seems less symmetric.

Curious to know whether you experimented with this design, and/or what your thoughts are about :arrow_up:

Reply:
> The setup seems to break symmetry between `M` and `N`. Because only `M` is required to be reflexive.
Yes, there is a slight symmetry breaking here. I expect you are aware but just to be clear, the asymmetry is only formal and note mathematical since `N` is also reflexive by [LinearEquiv.isReflexive_of_equiv_dual_of_isReflexive](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/PerfectPairing.html#LinearEquiv.isReflexive_of_equiv_dual_of_isReflexive).

> Can you prove `RootPairing.flip` which takes a root pairing on `(M, N)` and turns it into a root pairing on `(N, M)` by exchanging the roles of the roots and coroots?
Yes, the following works (e.g., directly below the lemma `root_coroot_two`):
```lean
def flip (_ : IsReflexive R N := e.isReflexive_of_equiv_dual_of_isReflexive) :
    RootPairing ι e.flip where
  root := P.coroot
  coroot := P.root
  coroot_root_two := P.root_coroot_two
  mapsTo_preReflection_root := P.mapsTo_preReflection_coroot
  mapsTo_preReflection_coroot := P.mapsTo_preReflection_root
```

> My intuition suggests to encode `e` as a bilinear map
> 
> ```
> e : M →ₗ[R] N →ₗ[R] R
> ```
> 
> so that you can use `e.flip` to swap the roles of `M` and `N`.
This is my intuition also. The reason I prefer `e : M ≃ₗ[R] N →ₗ[R] R` (together with the `[IsReflexive R M]` assumption) is that it means we can avoid defining a new type for perfect pairings. If we instead used `e : M →ₗ[R] N →ₗ[R] R` then we need the to store the hypotheses that the induced maps `M → Dual R N` and `N → Dual R M` are both bijective so we either have to carry these hypotheses round manually or introduce a new type, say `PerfectPairing R M N`, that bundles them. So I see three approaches to perfect pairings as follows:

1. Current proposal: `e : M ≃ₗ[R] N →ₗ[R] R` + `[IsReflexive R M]`: disadvantage that have asymmetry that `IsReflexive R N]` not automatic.
2. Use `e : M →ₗ[R] N →ₗ[R] R` and two unbundled hypotheses: disadvantage that not much information encoded in the types so we'll have less automation (e.g., simp lemmas firing) and more verbose code.
3. Bundle `e : M →ₗ[R] N →ₗ[R] R` with the two hypotheses into a new `PerfectPairing R M N` type: disadvantage that we need to write and maintain more API for this new type. Also though this is more symmetric, now neither `[IsReflexive R M]` nor `[IsReflexive R N]` is automatic (though we probably need them less).

I'm open to all of the above approaches. I convinced myself that 1 was the best but I think it's a bit subjective and I'm willing to try alternative designs.

> Of course you would then need to require the pairing to be perfect, and I think this could be done as one of the fields of `RootPairing`.
For emphasis (with apologies for repeating myself) I am already requiring the pairing to be perfect. It sounds like you're saying that we could avoid defining perfect pairings independently and essentially merge 3 into the definition of `RootPairing`. I think this is worse overall since (for other non-root-data stuff) we need to have a theory of perfect pairings.

> There is of course `LinearEquiv.flip`, but it just seems less symmetric.
Right, I'm using [LinearEquiv.flip](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/PerfectPairing.html#LinearEquiv.flip) and it does break the symmetry. As I said above, perhaps this is the wrong call.

Reply:
Hmm, it's possible that approach 3 is what we want. For example `LinearEquiv.flip_flip` can never be true by definition and so if `P : RootPairing ι e` then `P.flip.flip : RootPairing ι e.flip.flip` is not even the same type as `P`. So maybe we bite the bullet and write + maintain more API to keep the symmetry?

Reply:
Oh I guess you mean we could just push `e` to be a field inside `RootPairing`. Maybe that would be the best. I'm out of time for today but will think about this tomorrow.

Reply:
Yes, I would like to get good defeqs for double flips. Because I think that will help with developing other parts of the theory, is my hunch.
I don't mind assuming both `[IsReflexive R M]` and `[IsReflexive R N]` if that preserves some symmetry, even if it is a bit redundant. But I'm still leaning slightly towards a `PerfectPairing` type.

Anyway, we could also leave this question as a TODO in some comment. But a refactor now might be cheaper than in 3 months, say.

Reply:
I've just pushed https://github.com/leanprover-community/mathlib4/pull/8882/commits/10f78d3b459c6caa8befb82b0f8528d5f11bad9f which needs a little polish but is looking promising.